Feature/202 section 02 testability (#203)
eea-oasis/baseline-standard: core/baseline-core-v1.0-psd01.md

#### **[D4]**	
Each specific product, service, asset, or set thereof offered and delivered by Provider(s) to Requester(s) SHOULD have its specific T&C document.

*This would allow the fine-graining and consistent application of commercial State-Object-specific business rules and data.*

[[D4]](#d4) Testability: BPI logic ensures State-Object-specific buisness rules and data exist per T&C document. An element such as a product in the state object can be uniquely(sp) identified and can therefor point to a circuit and can be forced to adhere to constraints written in the T&C

### 2.4.2 Commercial Documents 

Commercial documents, a category of commercial State Objects, refer to the state of a specific product/service/asset or set thereof, which may or may not be modified from an original offering to meet the Requester requirements and includes operational and commercial details. A commercial document is an abstract construct representing mutual commitments based on a legally binding contract.

#### **[R4]**	
A commercial State Object to be transacted on MUST be based on a specific commercial document.

> [[R4]](#r4) Testability: Given is an [Example](#functional-terms-implemented-as-zero-knowledge-circuit) of a specific commercial document necessary for a Commercial State Object to be transacted.

Author: Rewrote to reflect this comment

#### **[R5]**  
A commercial document MUST be derived from a legally binding contract.

> [[R5]](#r5) Testability: Given is an [Example](#commerical-agreement-as-verifiable-credential) of a legally binding contract. (legally binding contract can be expressed as given in example)

Reviewer: ```suggestion
[[R5]](#r5) Testability:  A legally binding contract outside of a BPI can be translated into a commercial document on the BPI as given in this [example](#commerical-agreement-as-verifiable-credential) of a legally binding contract expressed as a W3C Verifiable Credential which can be tested following the W3C Verifiable Credential test suite approach.
```

#### **[R6]**	
A commercial document MUST be represented as an electronic record on a BPI between the counterparties.

> [[R6]](#r6) Testability: Given is an [Example](#commerical-agreement-as-verifiable-credential) of a legally binding contract.

Reviewer: ```suggestion
[[R6]](#r6) Testability: Since the testability statement in [[R5]](#r5) represents an electronic version of a commercial document and is testable as stated, this requirement is testable as well.
```

#### **[R24]**	
The transacting counterparties MUST agree on the business process rules which are represented in the business workflows and worksteps in the BPI. 

[[R24]](#r24) Testability: A BPI can ensure both counterparties agree on buisness rules and buisness data required to validate the agreement by, for example, utilizing a [payment term agreement](#functional-terms-implemented-as-zero-knowledge-circuit) between two parties.

#### **[R25]** 	
The transacting counterparties MUST validate the correctness of a State Object based on a state change against the transaction business logic in the applicable BPI workflow and workstep.

> [[R25]](#r25) Testability: a BPI can validate the correctness of a State Object based on a state change against transaction logic by the validation zero-knowledge proof in the BPI and ultimately on the CCSM. Reference back to [[R16]](#r16) to understand the validation of a zero-knowledge proof in a smart contract. 

Reviewer: ```suggestion
[[R25]](#r25) Testability: A BPI can validate the correctness of a State Object based on a state change against transaction logic by verifying a zero-knowledge proof of the state change in the BPI, and ultimately on the utilized CCSM. Reference back to [[R16]](#r16) to understand the verification of a zero-knowledge proof in a smart contract. 
```

#### **[R26]** 	
The transacting counterparties MUST generate a Proof of Correctness of a State Object based on a state change that can be validated against the BPI transaction business logic.

[[R26]](#r26) Testability: This can be accomplished by creating a [privacy package](https://github.com/eea-oasis/baseline/tree/main/core/privacy) with Zero-Knowledge Circuits generating Zero-Knowledge Proofs of the State changes of a State Object. 

#### **[R27]** 	
Any new state between counterparties MUST be recorded on the BPI between them.

[[R27]](#r27) Testability: This can be accomplished by creating a new entry in the storage of the BPI for the new State, such as in this [example](#Storing-an-Agreement-as-State-Object-in-Merkle-Tree), which can be queried. 

#### **[R28]** 	
Any transacting counterparty having received a Proof of Correctness of a state change MUST be able to validate that Proof of Correctness against the BPI between the counterparties.

> [[R28]](#r28) Testability: Including a ZK Proof, public input, and state commitement inside the message payload would enable counterparties to validate the Proof of Correctness.

Reviewer: ```suggestion
[[R28]](#r28) Testability: Including a zero-knowledge proof, public input, and state commitment, together with the prover scheme, and if required the common reference string, inside the message payload would enable counterparties to validate the Proof of Correctness on their own utilizing known libraries for the prover scheme such a PLONK or Groth16.
```

*The agreement on the governance entity, its rules, and its method of achieving interval synchronization consensus, as well as the definition of acceptable governance structures and their rules is beyond the scope of this document.*

[[D9]](#d9) Testability: This can acheived through the implementation of a new adapter program for each CCSM to add a CCSM into a BPI. Each adapter program will be different for each CCSM though all adapter programs will have a constructTx and a sendTransaction function, which was given in the example in [[R11]](#r11). 

#### **[R31]**	
The transacting counterparties MUST agree on which BPI is to be used.

[[R31]](#r31) Testability: In order for counterparties to transact on the same BPI instance, this requierment must have been fulfilled.  

### 2.8.3 External Applications

#### **[R32]**	
Application/s providing transaction functionality such as billing to counterparties, and are, therefore, external concerning the BPI, MUST be independent of any BPI.

*Note, this requirement is motivated by reducing the dependency of counterparty internal systems on the BPI and vice versa.*

> [[R32]](#r32) Testability: Ensuring this means counterparties only interact with the BPI utilizing the APIs as defined Baseline Protocol Standard, and not through integration of functionalities of external systems directly into a BPI, for example integrating an API of an ERP system into a BPI.

Reviewer: ```suggestion
[[R32]](#r32) Testability: In an implementation, counterparties only interact with the BPI utilizing the APIs as defined by the Baseline Protocol Standard, and not through the integration of functionalities of external systems directly into a BPI, for example integrating an API of an ERP system into a BPI.
```
